tests/suites/test_suite_pkcs1_v21.data

```diff
@@ -883,3 +883,107 @@ pkcs1_rsassa_pss_sign:1048:16:"0f39b79809516becc2e3481b6b47584aa2299bd2027ab8a30
+
+RSASSA-PSS Signature, RSA-1024, SHA-256, slen = olen-hlen-2
+depends_on:MBEDTLS_SHA256_C
+pkcs1_rsassa_pss_sign_ext:1024:16:"e5563b145db6ff5a16280d3e80eff02f181dbd03324ef247f596a4d4a7b8daa32b9934e3c7f4dcf6a3105462dec63839638618418b51db02693fabb4e6838725":16:"d2a4ec0fa2226cde82da77653b072cd098535d3e90ed4d7224dcb8cb8b9314768dc517e22d7c8fa13f253daa7465a79956098aa4cc3a6e35e8b1fcc4f97e774f":16:"bcb47b2e0dafcba81ff2a2b5cb115ca7e757184c9d72bcdcda707a146b3b4e29989ddc660bd694865b932b71ca24a335cf4d339c719183e6222e4c9ea6875acd528a49ba21863fe08147c3a47e41990b51a03f77d22137f8d74c43a5a45f4e9e18a2d15db051dc89385db9cf8374b63a8cc88113710e6d8179075b7dc79ee76b":16:"010001":MBEDTLS_MD_SHA256:MBEDTLS_MD_SHA256:"1248f62a4389f42f7b4bb131053d6c88a994db2075b912ccbe3ea7dc611714f14e075c104858f2f6e6cfd6abdedf015a821d03608bf4eba3169a6725ec422cd9069498b5515a9608ae7cc30e3d2ecfc1db6825f3e996ce9a5092926bc1cf61aa42d7f240e6f7aa0edb38bf81aa929d66bb5d890018088458720d72d569247b0c":"6f2841166a64471d4f0b8ed0dbb7db32161da13b3af3c4bd6d4cfcad8a03290e237b0cb3f05a4640d4ff655aa36fd36b4089817a7d4538ea9134971c37c12a5b3c360e2c90546c6553d2bff7419262821ce3fc99283483b9691ad5a0dbff":"":95:-16512
+
```

Comment:
There aren't enough boundary test cases. (And on the other hand I don't think it's useful to test all combinations of SHA-2 variants and 5 key sizes.) Missing:

* salt length = 0
* salt length = hash length (because it's the maximum value, except when the key is too small)
* salt length = the maximum for at least two different key sizes and at least two different hash lengths
* Copy and adapt the salt-related negative test cases for the old `sign` function: where an explicit salt would be correct, test at least the maximum possible salt length, and test that an explicit max+1 fails; where even an empty salt wouldn't fit, test that 0 fails.

Reply:
Thanks. I'll get on it over the week-end.

Reply:
I agree on all of the propose test cases except on the part where salt length == hash length. As far as I can tell, there is not hard limit on the salt length. As long as the length of the key allow to fit the salt. For instance if you have a 4096 bits key with SHA-256, you can have up to 4096/8 - 256/8 - 2 = 512-32-2 = 478 byte of salt. And in the PR code it is allowed in fact.

Did you maybe mean something else here? 

I'll add all other proposed test cases.

Reply:
[FIPS 186](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf) limits the salt length to the hash length:

> the length (in bytes) of the salt (sLen) shall satisfy 0 ≤ sLen ≤ hLen,

(with an additional constraint when sLen=hLen would overflow the key size). Using the NIST maximum is a very common choice (in fact, I think salt lengths other than 0 or max(salt length, max that fits) is rare).